src/runtime/contrib/cudnn/cudnn_json_runtime.cc

```diff
@@ -0,0 +1,198 @@
+        else
+          LOG(FATAL) << "Unsupported layout: " << layout;
+
+        if (attr_in_name(op_name, "relu")) {
+          act = CUDNN_ACTIVATION_RELU;
+        } else if (attr_in_name(op_name, "relu6")) {
+          act = CUDNN_ACTIVATION_CLIPPED_RELU;
+          coef = 6.0;
+        } else if (attr_in_name(op_name, "leaky_relu")) {
+          act = CUDNN_ACTIVATION_RELU;
+          coef = 0.1;
+        }
+        this->handle = entry_ptr->handle;
+        this->kernel_node = node;
+
+        // find best algo
+        TVMRetValue best_algo;
+
+        tvm::contrib::FindAlgo(format, dims, groups, padding.data(), strides.data(),
+                               dilation.data(), input_dims.data(), kernel_dims.data(),
+                               output_dims.data(), conv_dtype, conv_dtype, false, &best_algo);
+
+        this->algo = best_algo.operator int();
+      }
+    }
+  }
+
+  const char* type_key() const override { return "cudnn_json"; }  // May be overridden
+
+  void Run() override {
+    auto get_inputs = [this](const JSONGraphNode& node, bool has_bias) {
+      const DLTensor* bias = nullptr;
+      if (has_bias) {
+        bias = GetInput(node, 2);
+      }
+      return std::make_tuple(GetInput(node, 0), GetInput(node, 1), bias);
+    };
+
+    auto [a_ptr, b_ptr, bias_ptr] = get_inputs(kernel_node, has_bias);
+    uint32_t output_eid = EntryID(outputs_[0]);
+    auto out_ptr = data_entry_[output_eid];
+
+    if (this->has_bias) {
+      tvm::contrib::CallCudnnConvolutionBiasActivationForward(
+          this->handle, this->stream, this->mode, this->format, this->algo, this->dims,
+          this->groups, this->act, this->coef, this->padding.data(), this->strides.data(),
+          this->dilation.data(), a_ptr, b_ptr, out_ptr, bias_ptr, this->conv_dtype);
+    } else {
+      tvm::contrib::CallCudnnConvolutionForward(
+          this->handle, this->stream, this->mode, this->format, this->algo, this->dims,
+          this->groups, this->padding.data(), this->strides.data(), this->dilation.data(), a_ptr,
+          b_ptr, out_ptr, this->conv_dtype);
+    }
+  }
+
+ private:
+  const DLTensor* GetInput(const JSONGraphNode& node, const int idx) {
+    ICHECK_LT(idx, node.GetInputs().size());
+    auto eid = EntryID(node.GetInputs()[idx]);
+    ICHECK(eid < data_entry_.size());
+    return data_entry_[eid];
+  }
+  /*conv op name*/
+  std::string op_name;
+  /*conv mode: CUDNN_CROSS_CORRELATION by default*/
+  int mode = CUDNN_CROSS_CORRELATION;
+  /*algo: by default we select the implicit gemm algo, will be tuned in the initial pass.*/
+  int algo = CUDNN_CONVOLUTION_FWD_ALGO_IMPLICIT_GEMM;
+  /*if has bias*/
+  bool has_bias = false;
+  /*args for function call*/
+  int act = CUDNN_ACTIVATION_IDENTITY;
+  double coef = 1.0;
+  int format = CUDNN_TENSOR_NHWC;
+  int dims = 2;
+  int groups = 1;
+  std::vector<int> padding;
+  std::vector<int> strides;
+  std::vector<int> dilation;
+  std::string conv_dtype;
```

Comment:
Please clean unnecessary member variables.

Reply:
thanks masahi, but I've thoroughly double reviewed the member variables, all of them are necessary given our current design (we extract the conv-related attributes during complication, and when invoking Run() at the runtime we do not need to parse it again.)